Zendesk source to Zendesk target with AWS Comprehend sentiment analysis (#55)
triggermesh/bridges: bridges/comprehend-zendesk/README.md

# Zendesk Event Source to Zendesk Event Target Sentiment Analysis via AWS Comprehend

> **Zendesk Source -> [AWS Comprehend](https://aws.amazon.com/comprehend/) -> Zendesk Target**

Reviewer: This is redundant with the title and the description.
```suggestion
```

![thumbs](./assets/tenor.gif)

## Bridge Intention

This bridge is designed to complete the following event *flow*:

> * Recieve new Ticket Event's from Zendesk

Reviewer: ```suggestion
* Receive _New Ticket_ events from Zendesk.
```

> 

Reviewer: ```suggestion
```

> * Determine the Ticket's sentiment via [AWS Comprehend](https://aws.amazon.com/comprehend/)

Reviewer: ```suggestion
* Determine the Ticket's sentiment via [AWS Comprehend](https://aws.amazon.com/comprehend/).
```

> 

Reviewer: ```suggestion
```

> * Create a label for the new Ticket's baised on the outcome of [AWS Comprehend](https://aws.amazon.com/comprehend/)

Reviewer: ```suggestion
* Create a label for the Ticket based on the outcome of the [AWS Comprehend](https://aws.amazon.com/comprehend/) sentiment analysis.
```

    +--------+       +--------+       +---------+
    | source |-------> broker <- - - -| transf. |
    +--------+       +--------+       +---------+
                                           |
                           +---------------+
                           |
                     +-----v--+       +--------+
                     | broker <- - - -| target |
                     +--------+       +--------+

#### Prerequisites for Deployment

From [AWS](https://aws.amazon.com/):

* [AWS Access key ID](https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html)

* [AWS Secret Access key](https://docs.aws.amazon.com/general/latest/gr/aws-sec-cred-types.html)

* [Region](https://aws.amazon.com/premiumsupport/knowledge-center/vpc-find-availability-zone-options/#:~:text=To%20find%20which%20Availability%20Zones,options%20in%20the%20Region%20selector.)

* [Language](https://docs.aws.amazon.com/comprehend/latest/dg/supported-languages.html) - This will tell Comprehend what language to expect

From [Zendesk](https://www.zendesk.com/):

* [API Token](https://support.zendesk.com/hc/en-us/articles/226022787-Generating-a-new-API-token)

> * [subdomain](https://support.zendesk.com/hc/en-us/articles/221682747-Where-can-I-find-my-Zendesk-subdomain-)

Reviewer: IMO not a good place to describe this. Better have a real doc page for your source at docs.triggermesh.io (user-oriented, with pictures and a step-by-step approach), and simply link it from here.

Reviewer: Done in https://github.com/triggermesh/docs/pull/21. We should now remove this section

### Deployment

After populating the required fields in the `bridge.yaml` file located in the `Zendesk-Zendesk` folder you can deploy by executing the following command from within this directory.

>     kubectl -n <namespace> apply -f bridge.yaml

Reviewer: We really should avoid telling people to use `kubectl`. If people want to know how to create Bridges, they should also refer to our docs. We shouldn't be repeating that in every single bridge directory.

Reviewer 2: As an alternative, you could show how to create the bridge using the UI since there's a yaml section that the bridge can be pasted into.

Reviewer: That as well, but it belongs in the docs repo then, so that we have a single source of truth instead of 1 truth per bridge.

#### More Info

Both of the [Zendesk Event Source](https://github.com/triggermesh/knative-sources) and [Zenformation](https://github.com/bridges/bridges/comprehend-zendesk/zenformation) are open source!

> <!-- The Zendesk Target is not open source at the moment, but you can find its acomonying documentation in the main [Triggermesh Documentation](https://docs.triggermesh.io/) -->

Reviewer: Conflicting statements.

# Sentiment Analysis of new Zendesk tickets using via AWS Comprehend

![Thumb up](./assets/thumb-up.gif)

This bridge is designed to complete the following event *flow*:

1. Receive _New Ticket_ events from the Zendesk event source.
1. Determine the Ticket's sentiment via [AWS Comprehend][aws-comprehend] using a custom event transformation.
1. Label the new Zendesk Ticket using the Zendesk event target, based on the outcome of the AWS Comprehend sentiment
   analysis.

## Architecture overview

Instead of subscribing directly to the events generated by the Zendesk event source, the Zendesk event target subscribes
to events from a _second_ event broker. This event broker contains the digested events from the AWS Comprehend
transformation service, which is also responsible for moving events from one broker to the other.

```
+--------+       +--------+       +---------+     +-------+
| source |-------> broker <- - - -| transf. |<--->|  AWS  |
+--------+       +--------+       +---------+     +-------+
                                       |
                       +---------------+
                       |
                 +-----v--+       +--------+

>                  | broker <- - - -| target |

Reviewer 2: Very minor nit, but shouldn't the arrow go from broker --> target instead of broker <-- target?

Reviewer: I wanted to distinguish betweenn push (source->broker) and pull (broker--trigger/subscription(pull)->target). Not sure how to do that without bloating the  diagram too much.

Reviewer 2: Dunno. Think the breaking up of two different brokers helps at least differentiate between the two given the common linkage of the transformation. Although I expected the results of the transformation to be a push to the target instead of a pull.

Reviewer: The pull is just how a Subcription/Trigger works. It pulls from a Broker's channel, and pushes to the target.
I'll change it to avoid the confusion.

Reviewer: Done in e619b37

                 +--------+       +--------+
```

## Prerequisites

### Transformation service configuration

The AWS Comprehend transformation component requires the following parameters:

* [**AWS Access Key ID and Secret Access Key**][aws-creds]: credentials to authenticate API requests to AWS Comprehend.
* [**Region**][aws-region]: AWS region in which the transformation service interacts with the AWS Comprehend API.

> * [**Language**][comprehend-lang]: the language AWS Comprehend should expect in the body of the Zendesk Ticket.

Reviewer 2: Where is the expected language variable defined by AWS?  I know where it is in the code, and  I see `en` specified in the bridge, but should this reference what AWS is expecting?

Reviewer: Check the rendered Markdown, [comprehend-lang] is a real link to exactly what you're looking for.

Reviewer 2: Okay okay.

### Secrets

The event bridge reads several secrets, such as the credentials mentioned above, from the following Kubernetes Secret
objects:

* `zendesksource`: credentials for the Zendesk source
* `awscomprehend`: credentials for the Transformation service (AWS Comprehend sentiment analysis)
* `zendesktarget`: credentials for the Zendesk target

Those Secrets need to exist in the TriggerMesh user namespace for the bridge components to start. Please refer to the

> [secrets.yaml](./secrets.yaml) file for an example of Secret objects and their expected attributes.

Reviewer 2: I'm glad to see this documented. Although it does raise the concern if this is intended to be instantiated globally or per user/namespace? If the latter, then this must be re-evaluated.

Reviewer: I'm not following, there is no notion of "global" in TriggerMesh. Could you elaborate?

Reviewer 2: Think the way I'm reading this initially is the `triggermesh` user namespace vs Triggermesh user's namespace `cab`.

Reviewer: Ahh, I get it now. Changing it.

Reviewer: Addressed in c6de30a
